File: typescript/Linked Lists/remove_kth_last_node.ts
import { ListNode } from './ds';

/* Definition of ListNode:
class ListNode {
    val: number; next: ListNode | null;
    constructor(val: number){
        this.val = val;
        this.next = null;
    }
}*/

function removeKthLastNode(head: ListNode | null, k: number): ListNode | null {
    // A dummy node to ensure there's a node before 'head' in case we 
    // need to remove the head node.
    let dummy: ListNode | null = new ListNode(-1);
    dummy.next = head;
    let trailer: ListNode | null = dummy;
    let leader: ListNode | null = dummy;
    // Advance 'leader' k steps ahead.  
    for (let i = 0; i < k; i++){
        leader = leader.next;
        // If k is larger than the length of the linked list, no node 
        // needs to be removed.
        if (leader === null) 
            return head;
    }
    // Move 'leader' to the end of the linked list, keeping 'trailer'
    // k nodes behind.
    while (leader.next !== null){
        leader = leader.next;
        trailer = trailer!.next;
    }
    // Remove the kth node from the end.
    trailer!.next = trailer!.next!.next;
    return dummy.next;
}

File: csharp/Linked Lists/RemoveKthLastNode.cs
using ds.ListNode;

/*
	Definition of ListNode:
	class ListNode
	{
		public ListNode(int val, ListNode next = null)
		{
			Val = val;
			Next = next;
		}

		public int Val { get; set; }

		public ListNode Next { get; set; }
	}
*/

public class Solution
{
	public ListNode RemoveKthLastNode(ListNode head, int k)
	{
		// A dummy node to ensure there's a node before 'head' in case we need to remove the head node.
		ListNode dummy = new ListNode(-1);
		dummy.Next = head;

		ListNode leader = dummy;
		ListNode trailer = dummy;

		// Advance 'leader' k steps ahead.
		for (int i = 0; i < k; i++)
		{
			leader = leader.Next;
			
			// If k is larger than the length of the linked list, no node needs to be removed.
			if (leader == null)
				return head;
		}

		// Move 'leader' to the end of the linked list, keeping 'trailer' k nodes behind.
		while (leader.Next != null)
		{
			leader = leader.Next;
			trailer = trailer.Next;
		}

		// Remove the kth node from the end.
		trailer.Next = trailer.Next.Next;
	
		return dummy.Next;
	}
}
